File: kotlin/protocol/src/main/kotlin/com/roomlink/protocol/Pointer.kt
package com.roomlink.protocol

import kotlin.math.hypot
import kotlin.math.min

/**
 * A port of swift/Sources/RoomLinkProtocol/Pointer.swift. Behaviour is held to
 * the Swift side by protocol/transcripts.txt.
 *
 * The wire carries a **button mask**, not clicks. That is RFB's design and it
 * is the reason a dropped or reordered message heals: every message states the
 * whole truth, so the next one corrects the last. A desktop wants the edges, so
 * [PointerState] computes them on the host, where the previous state is known.
 * A finger has no buttons, so [TouchIntent] reads what the person meant from
 * time and distance on the viewer.
 */

/** What the host should be told, derived from what the viewer says is true now. Coordinates are in whatever space the caller feeds it. */
class PointerState {
    enum class Button { LEFT, RIGHT }

    sealed interface Action {
        /** Nothing is held; the pointer is just travelling. */
        data class Move(val at: Point) : Action
        /** A button is held and the pointer moved. Not a move — posting a move mid-drag does nothing at all. */
        data class Drag(val at: Point, val button: Button) : Action
        data class Down(val at: Point, val button: Button, val clicks: Int) : Action
        /** Carries the same count as the press it closes. */
        data class Up(val at: Point, val button: Button, val clicks: Int) : Action
    }

    companion object {
        const val LEFT: UByte = 1u
        const val RIGHT: UByte = 2u
        /** A second click only counts as a double if it lands near enough and soon enough. */
        const val DOUBLE_CLICK_INTERVAL = 0.5
        const val SLOP = 4.0

        private fun all(mask: UByte): List<Button> = buildList {
            if ((mask and LEFT) == LEFT) add(Button.LEFT)
            if ((mask and RIGHT) == RIGHT) add(Button.RIGHT)
        }
        private fun first(mask: UByte): Button? = all(mask).firstOrNull()
    }

    private var held: UByte = 0u
    private var at: Point? = null
    private var lastDown: Pair<Point, Double>? = null
    private var clicks = 1

    /** The viewer says: these buttons are down, and the pointer is here. */
    fun admit(buttons: UByte, point: Point, now: Double): List<Action> {
        val out = ArrayList<Action>()
        // The first message always counts as movement — the host pointer is
        // wherever the presenter left it, which is not where the finger is.
        val moved = at != point
        val pressed = (buttons.toInt() and held.toInt().inv() and 0xFF).toUByte()
        val released = (held.toInt() and buttons.toInt().inv() and 0xFF).toUByte()

        // Position first, so a press lands where the finger is. Whether that
        // movement is a move or a drag is decided by what was held *before*.
        if (moved) {
            val down = first(held)
            if (down != null) out.add(Action.Drag(point, down)) else out.add(Action.Move(point))
        }
        for (button in all(released)) {
            out.add(Action.Up(point, button, if (button == Button.LEFT) clicks else 1))
        }
        for (button in all(pressed)) {
            var count = 1
            if (button == Button.LEFT) {
                val ld = lastDown
                val near = ld != null && hypot(ld.first.x - point.x, ld.first.y - point.y) <= SLOP
                val soon = ld != null && now - ld.second <= DOUBLE_CLICK_INTERVAL
                clicks = if (near && soon) min(clicks + 1, 3) else 1
                lastDown = Pair(point, now)
                count = clicks
            }
            out.add(Action.Down(point, button, count))
        }

        held = buttons
        at = point
        return out
    }

    /** Whatever is still held, released where it stands. For a viewer that vanished mid-drag. */
    fun release(now: Double): List<Action> {
        val point = at
        if (held == 0u.toUByte() || point == null) return emptyList()
        return admit(0u, point, now)
    }
}

/**
 * A finger has no buttons and no hover, so what the person meant has to be read
 * from time and distance: a quick lift is a click, a hold is a right click, and
 * movement past a threshold is a drag. Positions are in the view's coordinates.
 */
class TouchIntent {
    data class Step(val buttons: UByte, val at: Point)

    enum class Phase { BEGAN, MOVED, ENDED, CANCELLED }

    companion object {
        /** Far enough that it was not just the hand settling on the glass. */
        const val DRAG_SLOP = 10.0
        const val HOLD_DELAY = 0.5
    }

    private var start: Point? = null
    private var touching = false
    private var dragging = false
    /** The hold already fired a right click; the lift owes nothing more. */
    private var spent = false

    /** True while a hold timer is worth having scheduled. */
    val awaitingHold: Boolean get() = touching && !dragging && !spent

    fun admit(phase: Phase, point: Point): List<Step> = when (phase) {
        Phase.BEGAN -> {
            start = point
            touching = true
            dragging = false
            spent = false
            // Send the pointer to the finger before anything is pressed.
            listOf(Step(0u, point))
        }
        Phase.MOVED -> {
            val s = start
            if (!touching || s == null) emptyList()
            else if (dragging) listOf(Step(PointerState.LEFT, point))
            else if (spent || hypot(point.x - s.x, point.y - s.y) <= DRAG_SLOP) listOf(Step(0u, point))
            else {
                dragging = true
                // Press at where the finger *started*: dragging a window has to
                // grab the titlebar that was touched.
                listOf(Step(PointerState.LEFT, s), Step(PointerState.LEFT, point))
            }
        }
        Phase.ENDED -> {
            if (!touching) emptyList()
            else {
                touching = false
                if (dragging) listOf(Step(0u, point))
                else if (spent) emptyList()
                else listOf(Step(PointerState.LEFT, point), Step(0u, point))
            }
        }
        Phase.CANCELLED -> {
            if (!touching) emptyList()
            else {
                touching = false
                if (dragging) listOf(Step(0u, point)) else emptyList()
            }
        }
    }

    /** The hold timer went off. Returns nothing if the finger already left or started dragging. */
    fun holdFired(): List<Step> {
        val s = start
        if (!awaitingHold || s == null) return emptyList()
        spent = true
        return listOf(Step(PointerState.RIGHT, s), Step(0u, s))
    }
}

File: swift/Sources/RoomLinkProtocol/Pointer.swift
import CoreGraphics
import Foundation


/// Turning a viewer's finger into a pointer on the presenter's Mac.
///
/// Two state machines face each other across the link, and neither one is
/// where the interesting failure lives — that is in the translation.
///
/// The wire carries a **button mask**, not clicks. That is RFB's design
/// (RFC 6143) and it is the reason a dropped or reordered message heals: every
/// message states the whole truth, so the next one corrects the last. Click
/// and release *events* cannot do that — lose a release and the button is
/// stuck down until somebody notices, which on a link measured losing a fifth
/// of what it sends is a matter of when, not whether.
///
/// macOS wants the edges, so `PointerState` computes them here, on the host,
/// where the previous state is known.

// MARK: - Host: mask -> events

/// What the presenter's Mac should be told, derived from what the viewer says
/// is true now. Coordinates are in whatever space the caller feeds it; the
/// host feeds global display points, so the slop below is points.
public struct PointerState {
    public enum Button: Equatable { case left, right }

    public enum Action: Equatable {
        /// Nothing is held; the pointer is just travelling.
        case move(CGPoint)
        /// A button is held and the pointer moved. This is *not* a move —
        /// posting `mouseMoved` mid-drag silently does nothing at all, which
        /// is the single easiest way to ship a broken drag.
        case drag(CGPoint, Button)
        case down(CGPoint, Button, clicks: Int)
        /// Carries the same count as the press it closes — AppKit controls
        /// read `clickCount` off the release as often as off the press.
        case up(CGPoint, Button, clicks: Int)
    }

    public static let left: UInt8 = 1
    public static let right: UInt8 = 2

    /// A second click only counts as a double if it lands near enough and soon
    /// enough. Both bounds are the system's, in spirit: ~0.5 s and a few
    /// points. Two clicks in the same place a second apart are two clicks.
    public static let doubleClickInterval: TimeInterval = 0.5
    public static let slop: CGFloat = 4

    private var held: UInt8 = 0
    private var at: CGPoint?
    private var lastDown: (at: CGPoint, when: TimeInterval)?
    private var clicks = 1

    /// The viewer says: these buttons are down, and the pointer is here.
    public init() {}

    public mutating func admit(buttons mask: UInt8, at point: CGPoint, now: TimeInterval) -> [Action] {
        var out: [Action] = []
        // The first message always counts as movement — the host pointer is
        // wherever the presenter left it, which is not where the finger is.
        let moved = at != point
        let pressed = mask & ~held
        let released = held & ~mask

        // Position first, so a press lands where the finger is rather than
        // where the pointer happened to be sitting. Whether that movement is a
        // move or a drag is decided by what was held *before* this message:
        // during a release the button is still down for the travel.
        if moved {
            if let down = Self.first(in: held) {
                out.append(.drag(point, down))
            } else {
                out.append(.move(point))
            }
        }
        for button in Self.all(in: released) {
            out.append(.up(point, button, clicks: button == .left ? clicks : 1))
        }
        for button in Self.all(in: pressed) {
            var count = 1
            if button == .left {
                let near = lastDown.map { hypot($0.at.x - point.x, $0.at.y - point.y) <= Self.slop } ?? false
                let soon = lastDown.map { now - $0.when <= Self.doubleClickInterval } ?? false
                clicks = near && soon ? min(clicks + 1, 3) : 1
                lastDown = (point, now)
                count = clicks
            }
            out.append(.down(point, button, clicks: count))
        }

        held = mask
        at = point
        return out
    }

    /// Whatever is still held, released where it stands. For a viewer that
    /// vanished mid-drag: without this the button stays down on the host.
    public mutating func release(now: TimeInterval) -> [Action] {
        guard held != 0, let point = at else { return [] }
        return admit(buttons: 0, at: point, now: now)
    }

    private static func all(in mask: UInt8) -> [Button] {
        var out: [Button] = []
        if mask & left == left { out.append(.left) }
        if mask & right == right { out.append(.right) }
        return out
    }

    private static func first(in mask: UInt8) -> Button? { all(in: mask).first }
}

// MARK: - Viewer: finger -> mask

/// A finger has no buttons and no hover, so what the person meant has to be
/// read from time and distance: a quick lift is a click, a hold is a right
/// click, and movement past a threshold is a drag.
///
/// Positions are in the view's coordinates; the caller normalizes before
/// sending. `slop` is therefore in points on the phone's glass.
public struct TouchIntent {
    public struct Step: Equatable {
        public let buttons: UInt8
        public let at: CGPoint
        public init(buttons: UInt8, at: CGPoint) { self.buttons = buttons; self.at = at }
    }

    /// Far enough that it was not just the hand settling on the glass.
    public static let dragSlop: CGFloat = 10
    public static let holdDelay: TimeInterval = 0.5

    public enum Phase { case began, moved, ended, cancelled }

    private var start: CGPoint?
    private var touching = false
    private var dragging = false
    /// The hold already fired a right click; the lift owes nothing more.
    private var spent = false

    /// True while a hold timer is worth having scheduled.
    public init() {}

    public var awaitingHold: Bool { touching && !dragging && !spent }

    public mutating func admit(_ phase: Phase, at point: CGPoint) -> [Step] {
        switch phase {
        case .began:
            start = point
            touching = true
            dragging = false
            spent = false
            // Send the pointer to the finger before anything is pressed, so
            // the presenter's screen shows where the tap is about to land.
            return [Step(buttons: 0, at: point)]

        case .moved:
            guard touching, let start else { return [] }
            if dragging { return [Step(buttons: PointerState.left, at: point)] }
            guard !spent, hypot(point.x - start.x, point.y - start.y) > Self.dragSlop else {
                return [Step(buttons: 0, at: point)]
            }
            dragging = true
            // Press at where the finger *started*, not where it is now:
            // dragging a window has to grab the titlebar that was touched.
            return [Step(buttons: PointerState.left, at: start),
                    Step(buttons: PointerState.left, at: point)]

        case .ended:
            guard touching else { return [] }
            touching = false
            if dragging { return [Step(buttons: 0, at: point)] }
            if spent { return [] }
            return [Step(buttons: PointerState.left, at: point),
                    Step(buttons: 0, at: point)]

        case .cancelled:
            guard touching else { return [] }
            touching = false
            return dragging ? [Step(buttons: 0, at: point)] : []
        }
    }

    /// The hold timer went off. Returns nothing if the finger already left or
    /// started dragging — a timer that outlived its gesture is harmless, which
    /// saves every caller from having to cancel it precisely.
    public mutating func holdFired() -> [Step] {
        guard awaitingHold, let start else { return [] }
        spent = true
        return [Step(buttons: PointerState.right, at: start),
                Step(buttons: 0, at: start)]
    }
}
